cmd/kubeadm/app/cmd/phases/upgrade/node/controlplane.go

```diff
@@ -82,6 +83,12 @@ func runControlPlane() func(c workflow.RunData) error {
 			return errors.Wrap(err, "failed to perform addons upgrade")
 		}
 
+		if features.Enabled(cfg.FeatureGates, features.ControlPlaneKubeletLocalMode) {
+			if err := upgrade.UpdateKubeletLocalMode(cfg, dryRun); err != nil {
+				return errors.Wrap(err, "failed to update kubelet local mode")
+			}
+		}
+
```

Comment:
Sorry for the late comment, but shouldn't this be moved to the `kubelet-config` phase?
Now is in the `kubeadm upgrade node control-plane` phase.
cc @neolit123 @brosner

Reply:
not sure, it's a feature relevant to the control plane. while kubelet-config is about the kubeletconfiguration. why is lubele-config a better place?

Reply:
My concern is that the `kubelet.conf` is part of the kubelet configuration, only used by kubelet component. Therefore, it is better for them to be together, if possible, in a phase. 
The `control-plane` phase focus only on things about the control plane components, such as etcd, kube-apiserver, kube-scheduler, kube-controller-manager and addons. 
 The `kubelet-config` phase focus only on kubelet, so users can use `kubeadm upgrade node --skip-phases kubelet-config` to skip the upgrade about kubelet. Then they can do some custom process about kubelet. 
Of cource, we should check if the node is a control-plane node before run `upgrade.UpdateKubeletLocalMode` in `kubelet-config` phase. If it is not a control-plane node, just skip.

WDYT? @neolit123

Reply:
i think this deservers a new phase:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase/

control-plane
> Upgrade the control plane instance deployed on this node, if any

kubelet-config
> Download the kubelet configuration from the kubelet-config ConfigMap stored in the cluster

setting the server in a kubeconfig does not fit in any of the existing phases.

IMO:
- we should add a new hidden phase called `kubeconfig` with a description:
> Make updates to kubeconfig files on the node during upgrade.
- the phase should be called after the control-plane phase and before the kubelet-config phase

this mutation of the kubeconfig server will existing only during the transition period when the FG is exists.
so once this mutation is done the phase will remain empty as far as i can tell. we can keep it hidden and make it unhidden if we need to add anything more concrete in the future.

Reply:
Got it. If this is only temporary, i'm ok with the current way.

Reply:
i think it's still worth it moving to a new hidden kubeconfig phase of upgrade node.
@SataQiu @chrischdi

Reply:
Opened https://github.com/kubernetes/kubernetes/pull/126098 with the implementation and for further discussion